src/audio/mux/mux_generic.c

```diff
@@ -18,347 +18,503 @@
+			lookup[0].copy_elem[elem].in_ch;
+		lookup[0].copy_elem[elem].src_inc = source->channels;
+
+		lookup[0].copy_elem[elem].dest = (int16_t *)sink->w_ptr +
+			lookup[0].copy_elem[elem].out_ch;
+		lookup[0].copy_elem[elem].dest_inc = sink->channels;
+	}
 }
 
-/* \brief Demuxing 24 bit streams.
- *
- * Source stream is routed to sink with regard to routing bitmasks from
- * mux_stream_data structure. Each bitmask describes composition of single
- * output channel.
+static void demux_init_look_up_pointers_s32(struct comp_dev *dev,
+					    struct audio_stream *sink,
+					    const struct audio_stream *source,
+					    struct mux_look_up *lookup)
+{
+	uint8_t elem;
+
+	/* init pointers */
+	for (elem = 0; elem < lookup[0].num_elems; elem++) {
+		lookup[0].copy_elem[elem].src = (int32_t *)source->r_ptr +
+			lookup[0].copy_elem[elem].in_ch;
+		lookup[0].copy_elem[elem].src_inc = source->channels;
+
+		lookup[0].copy_elem[elem].dest = (int32_t *)sink->w_ptr +
+			lookup[0].copy_elem[elem].out_ch;
+		lookup[0].copy_elem[elem].dest_inc = sink->channels;
+	}
+}
+
+#if CONFIG_FORMAT_S16LE
+/**
+ * Source stream are routed to sinks with regard to look up table based on
+ * routing bitmasks from mux_stream_data structures array. Each sink channel
+ * has it's own lookup[].copy_elem describing source and sink fragment of
+ * memory featured in copying.
  *
- * \param[in,out] sink Destination buffer.
- * \param[in,out] source Source buffer.
- * \param[in] frames Number of frames to process.
- * \param[in] data Parameters describing channel count and routing.
+ * @param[in] dev Component device
+ * @param[in,out] sink Destination buffer.
+ * @param[in,out] sources Array of source buffers.
+ * @param[in] frames Number of frames to process.
+ * @param[in] lookup mux look up table.
  */
-static void demux_s24le(struct audio_stream *sink,
+static void demux_s16le(struct comp_dev *dev, struct audio_stream *sink,
 			const struct audio_stream *source, uint32_t frames,
-			struct mux_stream_data *data)
+			struct mux_look_up *lookup)
 {
-	int32_t sample;
-	int32_t *dst;
-	uint32_t dst_idx;
 	uint8_t i;
-	uint8_t out_ch;
-
-	for (i = 0; i < frames; i++) {
-		for (out_ch = 0; out_ch < sink->channels; out_ch++) {
-			sample = calc_sample_s24le(source,
-						   i * source->channels,
-						   data->mask[out_ch]);
-
-			/* saturate to 24 bits */
-			dst_idx = i * sink->channels + out_ch;
-			dst = audio_stream_write_frag_s32(sink, dst_idx);
-			*dst = sat_int24(sample);
+	uint8_t elem;
+	int16_t *src;
+	int16_t *dst;
+	uint32_t frames_without_wrap;
+
+	comp_dbg(dev, "demux_s16le()");
+
+	demux_init_look_up_pointers_s16(dev, sink, source, lookup);
+
+	while (frames) {
+		frames_without_wrap = demux_calc_frames_without_wrap(dev, sink,
+								     source,
+								     lookup,
+								     frames);
+
+		frames_without_wrap = frames < frames_without_wrap ? frames :
+			frames_without_wrap;
+
+		for (i = 0; i < frames_without_wrap; i++) {
```

Comment:
ok, this looks more efficient. Btw, why do we always have lookup[0] ?

Reply:
@lgirdwood 
I've changed using of look up table to ```loookup->copy_elem``` (```lookup[0]``` could be confusing).  Basically, we invoke process function with one look up table (We have one look up table for each sink - mux has one look up table (one sink), demux can have several look ups (several sinks)). For mux we invoke process function once:
``` c
cd->mux(dev, &sink->stream, &sources_stream[0], frames, &cd->active_lookup);
```
For demux we invoke process several times (depends on sink amount) with corresponding look up table:
``` c
/* produce output, one sink at a time */
	for (i = 0; i < MUX_MAX_STREAMS; i++) {
		if (!sinks[i])
			continue;

		buffer_invalidate(source, source_bytes);
		cd->demux(dev, &sinks[i]->stream, &source->stream, frames,
			  look_ups[i]);
		buffer_writeback(sinks[i], sinks_bytes[i]);
	}
```